dagfactory/dagbuilder.py

```diff
@@ -549,52 +549,102 @@ def make_task_groups(task_groups: Dict[str, Any], dag: DAG) -> Dict[str, "TaskGr
-                        ):
-                            task_group_conf["default_args"]["on_failure_callback"]: Callable = import_string(
-                                task_group_conf["default_args"]["on_failure_callback"]
-                            )
+            if utils.check_dict_key(task_group_conf["default_args"], "on_failure_callback"):
+                if isinstance(
+                    task_group_conf["default_args"]["on_failure_callback"],
+                    str,
+                ):
+                    task_group_conf["default_args"]["on_failure_callback"]: Callable = import_string(
+                        task_group_conf["default_args"]["on_failure_callback"]
+                    )
 
-                    if utils.check_dict_key(task_group_conf["default_args"], "on_retry_callback"):
-                        if isinstance(
-                            task_group_conf["default_args"]["on_retry_callback"],
-                            str,
-                        ):
-                            task_group_conf["default_args"]["on_retry_callback"]: Callable = import_string(
-                                task_group_conf["default_args"]["on_retry_callback"]
-                            )
+            if utils.check_dict_key(task_group_conf["default_args"], "on_retry_callback"):
+                if isinstance(
+                    task_group_conf["default_args"]["on_retry_callback"],
+                    str,
+                ):
+                    task_group_conf["default_args"]["on_retry_callback"]: Callable = import_string(
+                        task_group_conf["default_args"]["on_retry_callback"]
+                    )
+        return task_group_conf
 
-                task_group = TaskGroup(**{k: v for k, v in task_group_conf.items() if k not in SYSTEM_PARAMS})
-                task_groups_dict[task_group.group_id] = task_group
-        return task_groups_dict
+    @staticmethod
+    def make_nested_task_groups(
+        task_group_name: str,
+        task_group_conf: Any,
+        task_groups_dict: Dict[str, "TaskGroup"],
+        task_groups: Dict[str, Any],
+        circularity_check_queue: List[str],
+        dag: DAG,
+    ):
+        """Takes a DAG and task group configurations. Creates nested TaskGroup instances.
+        :param task_group_name:
+        :param task_group_conf:
+        :param task_groups_dict
+        :param task_groups: Task group configuration from the YAML configuration file.
+        :param circularity_check_queue:
+        :param dag: DAG instance that task groups to be added.
+        """
+        if task_group_name in task_groups_dict:
+            return
+
+        if task_group_name in circularity_check_queue:
+            error_string = "Circular dependency detected:\n"
+            index = circularity_check_queue.index(task_group_name)
+            while index < len(circularity_check_queue):
+                error_string += f"{circularity_check_queue[index]} depends on\n"
```

Comment:
can we also append the task_group_name here itself instead of doing it after the while loop, wondering if there are multiple passes in the while iteration the string might appear incomplete.